src/devtool/evm.zig

```diff
@@ -245,91 +312,395 @@ pub fn stepExecute(self: *DevtoolEvm) !DebugStepResult {
+    while (!made_visible_progress) {
+        if (self.instr_index >= instructions.len) break;
+
+        const inst = instructions[self.instr_index];
+        switch (inst.arg) {
+            .block_info => |block| {
+                // Validate gas for whole block
+                if (frame.gas_remaining < block.gas_cost) {
+                    frame.gas_remaining = 0;
+                    exec_err = Evm.ExecutionError.Error.OutOfGas;
+                    self.is_completed = true;
+                    break;
+                }
+                frame.gas_remaining -= block.gas_cost;
+
+                // Validate stack bounds for block
+                const current_stack_size: u16 = @intCast(frame.stack.size());
+                if (current_stack_size < block.stack_req) {
+                    exec_err = Evm.ExecutionError.Error.StackUnderflow;
+                    self.is_completed = true;
+                    break;
+                }
+                if (current_stack_size + block.stack_max_growth > 1024) {
+                    exec_err = Evm.ExecutionError.Error.StackOverflow;
+                    self.is_completed = true;
+                    break;
+                }
+                // Advance past BEGINBLOCK and continue loop to reach a visible opcode
+                self.instr_index += 1;
+                continue;
+            },
+            .conditional_jump => |true_target| {
+                // Optimized JUMPI: destination resolved at analysis time
+                const cond = frame.stack.pop_unsafe();
+                if (cond != 0) {
+                    const base: [*]const @TypeOf(inst) = instructions.ptr;
+                    const idx = (@intFromPtr(true_target) - @intFromPtr(base)) / @sizeOf(@TypeOf(inst));
+                    self.instr_index = idx;
+                    // Map instruction index to PC for debugger view
+                    if (self.findPcForInstructionIndex(idx)) |pc_val| {
+                        new_pc = pc_val;
+                    }
+                } else {
+                    self.instr_index += 1;
+                    if (pc_before < self.bytecode.len) new_pc = pc_before + 1;
+                }
+                made_visible_progress = true;
+                continue;
+            },
+            .word => |value| {
+                // PUSH value onto stack (visible)
+                self.instr_index += 1;
+                frame.stack.append(value) catch |err| {
+                    exec_err = err;
+                };
+                // advance PC by PUSH opcode size (opcode + immediate length)
+                if (pc_before < self.bytecode.len) {
+                    const op = self.bytecode[pc_before];
+                    const imm_len: usize = if (op == 0x5f) 0 else if (op >= 0x60 and op <= 0x7f) @intCast(op - 0x5f) else 0;
+                    new_pc = pc_before + 1 + imm_len;
+                }
+                made_visible_progress = true;
+            },
```

Comment:
would be easier to instrument the existing code than copy paste new code like this